homeassistant/components/wemo/__init__.py

```diff
@@ -136,8 +140,12 @@ async def async_stop_wemo(_: Event | None = None) -> None:
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a wemo config entry."""
-    # This makes sure that `entry.async_on_unload` routines run correctly on unload
-    return True
+    await hass.data[DOMAIN]["stop"]()
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry, hass.data[DOMAIN]["platforms"]
+    )
+    hass.data[DOMAIN] = _domain_data(hass.data[DOMAIN].pop("config"))
```

Comment:
This function is about unloading a config entry, not about preparing to potentially load it again some time in the future.

In `async_setup` we should create an object in hass.data, in which the YAML config, and anything else which should persist between config entry reloads, is stored.
In `async_setup_entry` we should add that which is needed by the config entry.
In `async_unload_entry` we should remove that which was needed by the config entry.

Maybe something like this:
```
@dataclass
class WemoData:
    config: ConfigType
    config_entry_data: WemoConfigEntryData | None = None # Set by `async_setup_entry`, cleared by `async_unload_entry`
```

Reply:
Sounds good. I've done this and also changed some of the layering/composition too.

- `async_setup` is now responsible for initializing the `hass.data[DOMAIN]` as well as the pywemo SubscriptionRegistry & DiscoveryResponder. I've moved this pywemo initialization here as in the future it is likely that we'll want to deprecated the configuration.yaml settings and move to one Config Entry per device. When that happens the pywemo instances will need to be shared across all Config Entities.
- `async_setup_entry` & `async_unload_entry` now do as you suggest, adding and removing only what is needed for the config entry.
- I've moved everything to do with Platforms into the WemoDispatcher class. This also included simplifying a bit of the boilerplate code in each platform module (`async_wemo_dispatcher_connect`).
- I made a few helpers to assign the correct types from `hass.data[DOMAIN]`: `_async_coordinators`, `_async_registry`, and `_async_wemo_data`
- I overrode the `async_shutdown` method for the DataUpdateCoordinator to properly unsubscribe from push notifications.

Reply:
> Aside: I wish I could use _async_wemo_data from within wemo_device.py, but can't due to circular dependencies. These helpers & the asserts within seemed like a reasonable way to improve type safety and catch mistakes with tests.

I think you can move the new type definitions and the helper for getting the typed instance from `hass.data` to, for example, `wemo/models.py` to avoid the circular dependency?

Reply:
Yes, that and `if TYPE_CHECKING:` solve this nicely. Thanks for the suggestion.
Done in [cfde9bb](https://github.com/home-assistant/core/pull/96620/commits/cfde9bbe83ce428eed16a5bf1c590c4c764c8a02)